### PR TITLE
Internal: Add atomic controls [EDS-235]

### PR DESCRIPTION
**Pull request by Omerisra6**

### CODE REVIEW

File: includes/autoloader.php [hunk 1]
```diff
@@ -112,6 +112,8 @@ public static function get_classes_map() {
 	private static function init_classes_map() {
 		self::$classes_map = [
 			'Api' => 'includes/api.php',
+			'Atomic_Control' => 'includes/atomic-controls/atomic-control.php',
+			'Atomic_Section' => 'includes/atomic-controls/atomic-section.php',
```
Comment (Nevoss) [inline includes/autoloader.php:116]: Let's talk about moving all the classes into module

File: includes/widgets/atomic-heading.php [hunk 2]
```diff
@@ -28,6 +28,28 @@ protected function render() {
...
+					->set_options([
+						'h1' => 'H1',
+						'h2' => 'H2',
+						'h3' => 'H3',
+						'h4' => 'H4',
+						'h5' => 'H5',
+						'h6' => 'H6',
+					])
+					->get_props()
+				),
```
Comment (Nevoss) [inline includes/widgets/atomic-heading.php:45]: Let's rethink about `Select_Atomic_Control` and remove the `Select_Control`

File: includes/widgets/atomic-heading.php [hunk 3]
```diff
@@ -28,6 +28,28 @@ protected function render() {
...
+			Atomic_Control::bind_to( 'title' )
+				->set_label( __( 'Title', 'elementor' ) )
+				->set_type( Text_Control::KEY )
```
Comment (Nevoss) [inline includes/widgets/atomic-heading.php:48]: Is it `Text_Control` or `Textarea_Control`?

**Review by Nevoss — DISMISSED** [review]

File: includes/autoloader.php [hunk 4]
```diff
@@ -112,7 +112,6 @@ public static function get_classes_map() {
 	private static function init_classes_map() {
 		self::$classes_map = [
 			'Api' => 'includes/api.php',
-			'Atomic_Widget_Base' => 'includes/base/atomic-widget-base.php',
```
Comment (ronros-elementor) [inline includes/autoloader.php:115]: shouldn't it be on a separated PR?

**Review by ronros-elementor — CHANGES_REQUESTED** [review]

**Review by ronros-elementor — DISMISSED** [review]

File: modules/atomic-widgets/atomic-controls/types/textarea-control.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+namespace Elementor\Modules\AtomicWidgets\AtomicControls\Types;
+
+use Elementor\Modules\AtomicWidgets\Base\Atomic_Control_Base;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+class Textarea_Control extends Atomic_Control_Base {
+	private string $placeholder = '';
```
Comment (IshayMaya) [inline modules/atomic-widgets/atomic-controls/types/textarea-control.php:11]: maybe should be undefined if its empty?

File: modules/atomic-widgets/atomic-controls/types/select-control.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		return 'select';
+	}
+
+	public function set_options( array $options ): self {
+		$this->options = $options;
+
+		return $this;
+	}
+
+	public function get_options(): array {
```
Comment (Omerisra6) [inline modules/atomic-widgets/atomic-controls/types/select-control.php:24]: change to private

File: modules/atomic-widgets/atomic-controls/types/textarea-control.php (hunk 5 above)
Comment (Omerisra6) [reply modules/atomic-widgets/atomic-controls/types/textarea-control.php:11]: I will change this class to only hold the props and to set the placeholder in the props

File: modules/atomic-widgets/atomic-controls/atomic-section.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	public function get_items(): array {
+		return $this->value['items'];
+	}
+
+	public function jsonSerialize(): array {
+		return [
+			'type' => 'section',
+			'value' => $this->value,
```
Comment (StyleShit) [inline modules/atomic-widgets/atomic-controls/atomic-section.php:50]: "value" is a weird name
IDK what it means

File: modules/atomic-widgets/atomic-controls/atomic-section.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	public function set_items( array $items ): self {
+		$this->value['items'] = $items;
+
+		return $this;
+	}
+
+	public function get_label(): string {
+		return $this->value['label'];
```
Comment (StyleShit) [inline modules/atomic-widgets/atomic-controls/atomic-section.php:36]:
```suggestion
		return $this->value['label'] ?? null;
```

maybe? 
also others

File: modules/atomic-widgets/atomic-controls/types/select-control.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Elementor\Modules\AtomicWidgets\AtomicControls\Types;
```
Comment (StyleShit) [inline modules/atomic-widgets/atomic-controls/types/select-control.php:3]: do we really need to prefix everywhere?

```suggestion
namespace Elementor\Modules\AtomicWidgets\Controls\Types;
```

File: modules/atomic-widgets/widgets/atomic-heading.php [hunk 10]
```diff
@@ -1,9 +1,8 @@
 <?php
 namespace Elementor\Modules\AtomicWidgets\Widgets;
-
-use Elementor\Modules\AtomicWidgets\AtomicControls\Atomic_Control;
+use Elementor\Modules\AtomicWidgets\AtomicControls\Atomic_Section;
```
Comment (StyleShit) [inline modules/atomic-widgets/widgets/atomic-heading.php:3]: add an empty line

File: modules/atomic-widgets/widgets/atomic-heading.php [hunk 11]
```diff
@@ -36,6 +35,47 @@ protected function render() {
...
+//			->set_label( 'Tag' )
+//			->set_options( [
+//				[
+//					'value' => 'h1',
+//					'label' => 'H1',
+//				],
+//				[
+//					'value' => 'h2',
+//					'label' => 'H2',
+//				],
```
Comment (StyleShit) [inline modules/atomic-widgets/widgets/atomic-heading.php:48]: bruh

File: modules/atomic-widgets/widgets/atomic-heading.php
Comment (StyleShit) [inline modules/atomic-widgets/widgets/atomic-heading.php:1]: tests?

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/atomic-controls/atomic-section.php (hunk 7 above)
Comment (Omerisra6) [reply modules/atomic-widgets/atomic-controls/atomic-section.php:50]: My thoughts were that there are two types of panel items: Section and Control and they both hold two props: type (in our case: control or section) and each one of that panel items holds a value.
Maybe we can just make this object flat?

File: modules/atomic-widgets/atomic-controls/atomic-section.php (hunk 8 above)
Comment (Omerisra6) [reply modules/atomic-widgets/atomic-controls/atomic-section.php:36]: I think we should decide rather we want the section and control props to be accessible or not, if we don't want them to be accessible we could simply delete this method.

File: modules/atomic-widgets/widgets/atomic-heading.php
Comment (Omerisra6) [reply modules/atomic-widgets/widgets/atomic-heading.php:1]: No logic in the PR that we could test

File: modules/atomic-widgets/widgets/atomic-heading.php (hunk 10 above)
Comment (Omerisra6) [reply modules/atomic-widgets/widgets/atomic-heading.php]: Done.

File: modules/atomic-widgets/atomic-controls/types/select-control.php (hunk 9 above)
Comment (Omerisra6) [reply modules/atomic-widgets/atomic-controls/types/select-control.php]: Done.

File: modules/atomic-widgets/widgets/atomic-heading.php [hunk 12]
```diff
@@ -36,6 +36,48 @@ protected function render() {
...
+				],
+				[
+					'value' => 'h6',
+					'label' => 'H6',
+				],
+			]);
+
+		$title_control = Textarea_Control::bind_to( 'title' )
+			->set_label( 'Title' )
+			->set_placeholder( 'Hello, World!' );
```
Comment (StyleShit) [inline modules/atomic-widgets/widgets/atomic-heading.php:70]: translate

File: modules/atomic-widgets/widgets/atomic-heading.php [hunk 13]
```diff
@@ -36,6 +36,48 @@ protected function render() {
 	}
 
 	public function get_atomic_controls(): array {
-		return [];
+		$tag_control = Select_Control::bind_to( 'tag' )
+			->set_label( 'Tag' )
```
Comment (StyleShit) [inline modules/atomic-widgets/widgets/atomic-heading.php:40]: translate

File: modules/atomic-widgets/widgets/atomic-heading.php
Comment (StyleShit) [reply modules/atomic-widgets/widgets/atomic-heading.php:1]: shouldn't we test the infra somehow?

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/widgets/atomic-heading.php (hunk 12 above)
Comment (Omerisra6) [reply modules/atomic-widgets/widgets/atomic-heading.php:70]: Done.

File: modules/atomic-widgets/widgets/atomic-heading.php (hunk 13 above)
Comment (Omerisra6) [reply modules/atomic-widgets/widgets/atomic-heading.php:40]: Done.

**Review by StyleShit — APPROVED** [review]

**Pull request merged**